pkg/kubelet/rkt/log.go

```diff
@@ -0,0 +1,163 @@
+			glog.Warningf("rkt: Cannot unmarshal the journal log, skipping the line: %v", err)
+			continue
+		}
+
+		// Decode the json object as a map so we don't have to create a struct for it.
+		m := data.(map[string]interface{})
+		msg := m["MESSAGE"].(string)
+		tss := m["__REALTIME_TIMESTAMP"].(string)
+
+		ts, err := strconv.ParseInt(tss, 0, 64)
+		if err != nil {
+			glog.Warningf("rkt: Cannot parse the timestamp of the journal log, skipping the line: %v", err)
+			continue
+		}
+
+		// '_REALTIME_TIMESTAMP' is the microseconds since epoch.
+		// http://www.freedesktop.org/software/systemd/man/sd_journal_get_realtime_usec.html#Description
+		micros := time.Duration(ts) * time.Microsecond
+		t := time.Unix(int64(micros.Seconds()), 0)
+
+		var result string
+		if logOptions.Timestamps {
+			// Use the same time format as docker.
+			result = fmt.Sprintf("%s %s\n", t, msg)
+		} else {
+			result = fmt.Sprintf("%s\n", msg)
+		}
+
+		// When user interrupts the 'kubectl logs $POD -f' with 'Ctrl-C', this write will fail.
+		// We need to close the reader to force the journalctl process to exit, also we need to
+		// return here to avoid goroutine leak.
+		if _, err := io.WriteString(w, result); err != nil {
+			glog.Warningf("rkt: Cannot write the log to output: %v, data: %s", err, result)
+			return
+		}
+	}
+
+	if err := scanner.Err(); err != nil {
+		glog.Warningf("rkt: Cannot read journal logs", err)
+	}
+}
+
+// GetContainerLogs uses journalctl to get the logs of the container.
+// By default, it returns a snapshot of the container log. Set |follow| to true to
+// stream the log. Set |follow| to false and specify the number of lines (e.g.
+// "100" or "all") to tail the log.
+//
+// In rkt runtime's implementation, per container log is get via 'journalctl -m _HOSTNAME=[rkt-$UUID] -u [APP_NAME]'.
+// See https://github.com/coreos/rkt/blob/master/Documentation/commands.md#logging for more details.
+//
+// TODO(yifan): If the rkt is using lkvm as the stage1 image, then this function will fail.
+func (r *runtime) GetContainerLogs(pod *api.Pod, containerID string, logOptions *api.PodLogOptions, stdout, stderr io.Writer) error {
+	id, err := parseContainerID(containerID)
+	if err != nil {
+		return err
+	}
+
+	cmd := exec.Command("journalctl", "-m", fmt.Sprintf("_HOSTNAME=rkt-%s", id.uuid), "-u", id.appName, "-a")
+
+	// Get the json structured logs.
+	cmd.Args = append(cmd.Args, "-o", "json")
+
+	if logOptions.Follow {
+		cmd.Args = append(cmd.Args, "-f")
+	}
+
+	if logOptions.TailLines != nil {
+		cmd.Args = append(cmd.Args, "-n", strconv.FormatInt(*logOptions.TailLines, 10))
+	}
+
+	var since int64
+	if logOptions.SinceSeconds != nil {
+		t := unversioned.Now().Add(-time.Duration(*logOptions.SinceSeconds) * time.Second)
+		since = t.Unix()
+	}
+	if logOptions.SinceTime != nil {
+		since = logOptions.SinceTime.Unix()
+	}
+
+	cmd.Args = append(cmd.Args, "--since", time.Unix(since, 0).Format(journalSinceLayout))
```

Comment:
Nope, but I assumed `--since=0` is ok, however then I found https://github.com/systemd/systemd/issues/1477